### PR TITLE
New report router and user reporting refactoring

### PR DESCRIPTION
**Pull request by harshilsharma63**
#### Summary
* Created new admin report router and moved user reporting API into it.
* Refactored some interfaces for report filter and simplified the filter validation logic to be consistent with others and be more reusable.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-56189

#### Release Note
```release-note
none
```

### CODE REVIEW

File: server/channels/api4/report.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+)
+
+func (api *API) InitReports() {
+	api.BaseRoutes.Reports.Handle("/users", api.APISessionRequired(getUsersForReporting)).Methods("GET")
+}
+
+func getUsersForReporting(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !(c.IsSystemAdmin() && c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionSysconsoleReadUserManagementUsers)) {
```
Comment (larkox) [inline server/channels/api4/report.go:21]: 0/5 should this be and `&&` or an `||`. My guess is that sysadmins have always permissions for everything, so an `&&` here would be redundant. But I may be missing something regarding the permission schemes.

File: server/scripts/vet-api-check.sh [hunk 2]
```diff
@@ -55,6 +55,7 @@ OUTPUT_EXCLUDING_IGNORED=$(echo "$OUTPUT" | grep -Fv \
     -e 'Cannot find /api/v4/users/login/desktop_token method: POST in OpenAPI 3 spec.' \
     -e 'Cannot find /api/v4/users/notify-admin method: POST in OpenAPI 3 spec.' \
     -e 'Cannot find /api/v4/users/trigger-notify-admin-posts method: POST in OpenAPI 3 spec.' \
+    -e 'Cannot find /api/v4/reports/users method: GET in OpenAPI 3 spec.' \
```
Comment (larkox) [inline server/scripts/vet-api-check.sh:58]: Why is this needed? Somehow feels the wrong place to add this, but I don't understand 100% what is going on on this file :sweat_smile:

**Review by larkox — APPROVED** [review]
LGTM, just a couple of comments that don't really have to do with the refactoring.

File: server/channels/api4/report.go (hunk 1 above)
Comment (harshilsharma63) [reply server/channels/api4/report.go:21]: It's just rearranging existing code, but I think it's okay to explicitly mention what access we need. thoughts @devinbinnie ?

File: server/scripts/vet-api-check.sh (hunk 2 above)
Comment (harshilsharma63) [reply server/scripts/vet-api-check.sh:58]: Its mentioned in the PR where this was added - https://github.com/mattermost/mattermost/pull/23974

Pasting here for reference -

```
The underlying mattermost-govet tool effectively hasn't been called for some time, as we weren't checking out and building the spec files to pass. Now that the API is in the monorepo, build it locally and check against it.

Unfortunately, our API documentation isn't up-to-date, and this PR isn't fixing that. For now, add a discrete make vet-api and workflow that won't block the build until the API documentation is back in sync and can be merged into the existing make vet directive. Additionally, suppress errors from "known mismatches" (while still logging them), so we can at least start checking the API moving forward.
```

File: server/channels/api4/report.go (hunk 1 above)
Comment (devinbinnie) [reply server/channels/api4/report.go:21]: I'm not sure how this works explicitly, but I was under the impression that you can have an admin role without that permission, so we want to check for it. In any case I think most system admins do have it.

File: server/scripts/vet-api-check.sh (hunk 2 above)
Comment (devinbinnie) [reply server/scripts/vet-api-check.sh:58]: @harshilsharma63 This is only for existing API stuff that was added before the check was restored. This API actually is documented [here](https://github.com/mattermost/mattermost/blob/3e16d621a09e63c9611508ae91d79665e3568ba2/api/v4/source/users.yaml#L3254) so this needs to be moved over and this check needs to be enabled.

Comment (devinbinnie) [reply server/scripts/vet-api-check.sh:58]: ...which it looks like you did. So I don't know why this comment is needed.

File: server/public/model/channel.go [hunk 3]
```diff
@@ -35,6 +35,10 @@ const (
 
 	ChannelSortByUsername = "username"
 	ChannelSortByStatus   = "status"
+
+	ChannelReportingSortByDisplayName = "displayName"
```
Comment (devinbinnie) [inline server/public/model/channel.go:39]: Should these not go into `report.go`?

File: server/public/model/report.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		startAt = now.AddDate(0, -6, -0).UnixMilli()
+	}
+
+	options.StartAt = startAt
+	options.EndAt = endAt
+}
+
+func (options *ReportingBaseOptions) IsValid() *AppError {
+	// Don't allow fetching more than 100 users at a time from the normal query endpoint
+	if options.PageSize <= 0 || options.PageSize > ReportingMaxPageSize {
```
Comment (devinbinnie) [inline server/public/model/report.go:57]: These checks aren't really a property of the model itself, they more exist to protect the API. I originally had them in the `app` module but I'm starting to think they just belong in the API itself. So I think we should move them over there.

File: server/public/model/report.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	HideInactive bool
+}
+
+func (u *UserReportOptions) IsValid() *AppError {
+	if appErr := u.ReportingBaseOptions.IsValid(); appErr != nil {
+		return appErr
+	}
+
+	// Validate against the columns we allow sorting for
+	if !pUtils.Contains(UserReportSortColumns, u.SortColumn) {
```
Comment (devinbinnie) [inline server/public/model/report.go:100]: This check however I think is fine to keep here :)

**Review by devinbinnie — CHANGES_REQUESTED** [review]

File: server/scripts/vet-api-check.sh (hunk 2 above)
Comment (harshilsharma63) [reply server/scripts/vet-api-check.sh]: Fixed! So I had added a new `report.yaml` file but the code that consumes those yaml files needs to be told the names of all files, and doesn't automatically consume all yaml files in the directory 🤦‍♂️

Comment (harshilsharma63) [reply server/scripts/vet-api-check.sh]: So telling it the new file's name fixed it.

File: server/public/model/channel.go (hunk 3 above)
Comment (harshilsharma63) [reply server/public/model/channel.go]: Actually, channel reporting stuff shouldn't even be in this PR. I've removed this now.

File: server/public/model/report.go (hunk 4 above)
Comment (harshilsharma63) [reply server/public/model/report.go:55]: Makes sense. Done.

File: server/public/model/report.go (hunk 5 above)
Comment (harshilsharma63) [reply server/public/model/report.go:98]: Devin approved!

**Review by larkox — APPROVED** [review]

**Comment by larkox** [comment]
@harshilsharma63 Heads up that something seems "wrong" in the CI.

**Comment by harshilsharma63** [comment]
> @harshilsharma63 Heads up that something seems "wrong" in the CI.

Yes, its a test. Will be pushing a fix.

**Review by devinbinnie — APPROVED** [review]
Thanks @harshilsharma63! :)
